File: src/coep.ts
/**
 * coep.ts — Cross-Origin-Embedder-Policy response-header audit.
 * T76.
 *
 * COEP, paired with COOP=`same-origin`, enables `crossOriginIsolated`
 * document state — the modern browser primitive that gates
 * `SharedArrayBuffer`, high-resolution `performance.now()`,
 * and `performance.measureUserAgentSpecificMemory()`. Without
 * cross-origin isolation, Spectre-class side-channel attacks
 * can leak data from co-tenant origins inside the same browser
 * process.
 *
 * COEP also restricts WHICH cross-origin sub-resources the
 * browser will load: with `require-corp`, every cross-origin
 * fetched asset must explicitly opt-in via Cross-Origin-Resource-
 * Policy. With `credentialless`, cross-origin requests are
 * allowed but credentials are stripped (cookies, client certs,
 * basic-auth — the resource is fetched as if anonymous).
 *
 * Acceptable values:
 *   * `require-corp`     — strictest. All cross-origin
 *                          embeds need CORP.
 *   * `credentialless`   — permissive but credential-stripped;
 *                          newer (Chrome 96+, Firefox 119+).
 *   * `unsafe-none`      — explicit opt-out of isolation.
 *                          Browser default if header absent.
 *
 * Findings:
 *
 *   - coep.missing       warn   No header at all → defaults to
 *                               `unsafe-none`. Cross-origin
 *                               isolation cannot be enabled.
 *                               SharedArrayBuffer and high-res
 *                               timers stay disabled.
 *   - coep.unsafe-none   warn   Header explicitly set to
 *                               `unsafe-none`. Operator made the
 *                               choice — surfaced for confirmation.
 *   - coep.invalid       warn   Value not in the recognised set.
 *                               Browsers ignore (effective
 *                               `unsafe-none`).
 *
 * Out of scope: localhost (same family). http pages still warn
 * for missing — the cost of setting it is zero and the signal
 * is useful even before TLS is enforced.
 *
 * Reads from the same `topLevelResponseHeaders` Map as the
 * other response-header detectors. EIGHTH consumer of the
 * shared capture path.
 */

export interface CoepFinding {
  severity: 'strict' | 'warn';
  kind: string;
  detail: string;
  evidence: Record<string, unknown>;
}

export interface CoepSnapshot {
  pageUrl: string;
  pageIsLocalhost: boolean;
  raw: string | null;
}

const ACCEPTABLE: ReadonlySet<string> = new Set([
  'require-corp',
  'credentialless',
]);

const RECOGNISED: ReadonlySet<string> = new Set([
  ...ACCEPTABLE,
  'unsafe-none',
]);

function isLocalhost(url: string): boolean {
  try {
    const u = new URL(url);
    const h = u.hostname;
    return h === 'localhost' || h === '127.0.0.1' || h === '::1' || h.endsWith('.localhost');
  } catch {
    return false;
  }
}

function getHeader(headers: Record<string, string> | undefined, name: string): string | null {
  if (!headers) return null;
  for (const [k, v] of Object.entries(headers)) {
    if (k.toLowerCase() === name) return v;
  }
  return null;
}

export function buildCoepSnapshot(
  pageUrl: string,
  headers: Record<string, string> | undefined,
): CoepSnapshot {
  const pageIsLocalhost = isLocalhost(pageUrl);
  const raw = getHeader(headers, 'cross-origin-embedder-policy');
  return {
    pageUrl,
    pageIsLocalhost,
    raw: raw === null ? null : raw.trim().toLowerCase(),
  };
}

export function detectCoepIssues(snap: CoepSnapshot): CoepFinding[] {
  if (snap.pageIsLocalhost) return [];
  if (snap.raw === null) {
    return [{
      severity: 'warn',
      kind: 'coep.missing',
      detail: `No Cross-Origin-Embedder-Policy header. Defaults to 'unsafe-none' — cross-origin isolation cannot be enabled, so SharedArrayBuffer + high-resolution timers stay disabled and Spectre-class side-channel mitigations are unavailable. Set to 'require-corp' (strictest, requires every cross-origin sub-resource to opt in via CORP) or 'credentialless' (newer; allows cross-origin embeds without credentials).`,
      evidence: {},
    }];
  }
  if (snap.raw === 'unsafe-none') {
    return [{
      severity: 'warn',
      kind: 'coep.unsafe-none',
      detail: `Cross-Origin-Embedder-Policy explicitly set to 'unsafe-none'. Cross-origin isolation is disabled. Confirm this is intentional — the supersociety baseline is 'require-corp'.`,
      evidence: { value: snap.raw },
    }];
  }
  if (!RECOGNISED.has(snap.raw)) {
    return [{
      severity: 'warn',
      kind: 'coep.invalid',
      detail: `Cross-Origin-Embedder-Policy value '${snap.raw}' is not in the W3C-recognised set ('require-corp', 'credentialless', 'unsafe-none'). Browsers ignore unknown values and fall back to 'unsafe-none'.`,
      evidence: { value: snap.raw },
    }];
  }
  return [];
}

File: crates/crawler-detectors/src/coep.rs
//! `coep` — Cross-Origin-Embedder-Policy header detector.
//!
//! Mirror of `src/coep.ts`. Findings:
//!
//!   * `coep.missing`     warn  — header absent → defaults to unsafe-none
//!   * `coep.unsafe-none` warn  — explicit opt-out of isolation
//!   * `coep.invalid`     warn  — unrecognised value → effective unsafe-none
//!
//! COEP, paired with COOP=`same-origin`, enables the
//! `crossOriginIsolated` document state — the modern primitive
//! that gates `SharedArrayBuffer` and high-resolution timers
//! used to mitigate Spectre-class side channels.
//!
//! Acceptable values: `require-corp`, `credentialless`.
//! Recognised but discouraged: `unsafe-none`.
//!
//! AVP-2 invariants: `unsafe_code = "deny"`, pure detector, no I/O.

use crate::url_helpers::is_localhost;
use crate::{AxisFinding, AxisSeverity};
use serde::{Deserialize, Serialize};

/// Captured page state the detector consumes.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct CoepSnapshot {
    /// Page URL (used for localhost exemption).
    pub page_url: String,
    /// Localhost / loopback exemption.
    pub page_is_localhost: bool,
    /// Raw header value, lowercased + trimmed, or `None` if absent.
    pub raw: Option<String>,
}

const ACCEPTABLE: &[&str] = &["require-corp", "credentialless"];
const RECOGNISED: &[&str] = &["require-corp", "credentialless", "unsafe-none"];

fn lookup_header<I, K, V>(headers: I, name_lower: &str) -> Option<String>
where
    I: IntoIterator<Item = (K, V)>,
    K: AsRef<str>,
    V: AsRef<str>,
{
    for (k, v) in headers {
        if k.as_ref().to_ascii_lowercase() == name_lower {
            return Some(v.as_ref().to_owned());
        }
    }
    None
}

/// Build a snapshot from a page URL + a header map.
pub fn build_coep_snapshot<I, K, V>(page_url: &str, headers: I) -> CoepSnapshot
where
    I: IntoIterator<Item = (K, V)>,
    K: AsRef<str>,
    V: AsRef<str>,
{
    let raw = lookup_header(headers, "cross-origin-embedder-policy")
        .map(|s| s.trim().to_ascii_lowercase());
    CoepSnapshot {
        page_url: page_url.to_owned(),
        page_is_localhost: is_localhost(page_url),
        raw,
    }
}

/// Run the detector. Returns zero findings on localhost / loopback.
pub fn detect_coep_issues(snap: &CoepSnapshot) -> Vec<AxisFinding> {
    if snap.page_is_localhost {
        return Vec::new();
    }
    let Some(value) = snap.raw.as_deref() else {
        return vec![AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "coep.missing".into(),
            detail: "No Cross-Origin-Embedder-Policy header. Defaults to 'unsafe-none' — \
                cross-origin isolation cannot be enabled, so SharedArrayBuffer + \
                high-resolution timers stay disabled and Spectre-class side-channel \
                mitigations are unavailable. Set to 'require-corp' (strictest, requires \
                every cross-origin sub-resource to opt in via CORP) or 'credentialless' \
                (newer; allows cross-origin embeds without credentials)."
                .into(),
        }];
    };
    if value == "unsafe-none" {
        return vec![AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "coep.unsafe-none".into(),
            detail: format!(
                "Cross-Origin-Embedder-Policy explicitly set to 'unsafe-none'. \
                 Cross-origin isolation is disabled. Confirm this is intentional — \
                 the supersociety baseline is 'require-corp'. (value={value})"
            ),
        }];
    }
    if !RECOGNISED.iter().any(|r| *r == value) {
        return vec![AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "coep.invalid".into(),
            detail: format!(
                "Cross-Origin-Embedder-Policy value '{value}' is not in the W3C-recognised \
                 set ('require-corp', 'credentialless', 'unsafe-none'). Browsers ignore \
                 unknown values and fall back to 'unsafe-none'."
            ),
        }];
    }
    debug_assert!(ACCEPTABLE.iter().any(|a| *a == value));
    Vec::new()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn snap(url: &str, raw: Option<&str>) -> CoepSnapshot {
        let headers: Vec<(&str, &str)> = match raw {
            Some(v) => vec![("cross-origin-embedder-policy", v)],
            None => vec![],
        };
        build_coep_snapshot(url, headers)
    }

    #[test]
    fn localhost_exempt() {
        assert!(detect_coep_issues(&snap("https://localhost/", None)).is_empty());
        assert!(detect_coep_issues(&snap("http://127.0.0.1:8080/", None)).is_empty());
    }

    #[test]
    fn missing_header_warn() {
        let f = detect_coep_issues(&snap("https://example.com/", None));
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "coep.missing");
        assert_eq!(f[0].severity, AxisSeverity::Warn);
    }

    #[test]
    fn require_corp_no_finding() {
        let f = detect_coep_issues(&snap("https://example.com/", Some("require-corp")));
        assert!(f.is_empty(), "got {f:?}");
    }

    #[test]
    fn credentialless_no_finding() {
        let f = detect_coep_issues(&snap("https://example.com/", Some("credentialless")));
        assert!(f.is_empty(), "got {f:?}");
    }

    #[test]
    fn unsafe_none_explicit_warn() {
        let f = detect_coep_issues(&snap("https://example.com/", Some("unsafe-none")));
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "coep.unsafe-none");
    }

    #[test]
    fn unrecognised_value_warn() {
        let f = detect_coep_issues(&snap("https://example.com/", Some("banana")));
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "coep.invalid");
    }

    #[test]
    fn case_insensitive_value() {
        let f = detect_coep_issues(&snap("https://example.com/", Some("REQUIRE-CORP")));
        assert!(f.is_empty(), "got {f:?}");
    }

    #[test]
    fn whitespace_trimmed() {
        let f = detect_coep_issues(&snap("https://example.com/", Some(" require-corp ")));
        assert!(f.is_empty(), "got {f:?}");
    }

    #[test]
    fn header_lookup_case_insensitive() {
        let s = build_coep_snapshot(
            "https://example.com/",
            [("Cross-Origin-Embedder-Policy", "require-corp")],
        );
        assert!(detect_coep_issues(&s).is_empty());
    }

    #[test]
    fn http_page_still_warned() {
        let f = detect_coep_issues(&snap("http://example.com/", None));
        assert_eq!(f.len(), 1, "http pages still warned (cost-of-fix is zero)");
    }

    #[test]
    fn dot_localhost_subdomain_exempt() {
        assert!(detect_coep_issues(&snap("https://app.localhost/", None)).is_empty());
    }
}
